source/common/network/utility.cc

```diff
@@ -88,24 +88,68 @@ Api::IoCallUint64Result receiveMessage(uint64_t max_rx_datagram_size, Buffer::In
   return result;
 }
 
-} // namespace
-
-Address::InstanceConstSharedPtr Utility::parseInternetAddressNoThrow(const std::string& ip_address,
-                                                                     uint16_t port, bool v6only) {
+StatusOr<sockaddr_in> parseV4Address(const std::string& ip_address, uint16_t port) {
   sockaddr_in sa4;
   memset(&sa4, 0, sizeof(sa4));
-  if (inet_pton(AF_INET, ip_address.c_str(), &sa4.sin_addr) == 1) {
-    sa4.sin_family = AF_INET;
-    sa4.sin_port = htons(port);
-    return instanceOrNull(Address::InstanceFactory::createInstancePtr<Address::Ipv4Instance>(&sa4));
+  if (inet_pton(AF_INET, ip_address.c_str(), &sa4.sin_addr) != 1) {
+    return absl::FailedPreconditionError("failed parsing ipv4");
   }
+  sa4.sin_family = AF_INET;
+  sa4.sin_port = htons(port);
+  return sa4;
+}
+
+StatusOr<sockaddr_in6> parseV6Address(const std::string& ip_address, uint16_t port) {
   sockaddr_in6 sa6;
   memset(&sa6, 0, sizeof(sa6));
-  if (inet_pton(AF_INET6, ip_address.c_str(), &sa6.sin6_addr) == 1) {
+  const auto scope_pos = ip_address.rfind('%');
+  // TODO(#23952): Even though it would be nice to do any IPv6 parsing only with the getaddrinfo at
+  // the moment Windows parsing is slightly different in behavior than other platforms. For this
+  // reason we use the inet_pton for any parsing that does not contain the zone id.
+  if (scope_pos == std::string::npos) {
+    // Parse IPv6 with no scope.
+    if (inet_pton(AF_INET6, ip_address.c_str(), &sa6.sin6_addr) != 1) {
+      return absl::FailedPreconditionError("failed parsing ipv6");
+    }
     sa6.sin6_family = AF_INET6;
-    sa6.sin6_port = htons(port);
+  } else {
+    // Parse IPv6 with optional scope using getaddrinfo().
+    struct addrinfo hints;
+    memset(&hints, 0, sizeof(hints));
+    struct addrinfo* res = nullptr;
+    // Suppresses any potentially lengthy network host address lookups and inhibit the invocation of
+    // a name resolution service.
+    hints.ai_flags = AI_NUMERICHOST | AI_NUMERICSERV;
+    hints.ai_family = AF_INET6;
+    // Hint that getaddrinfo() need not return a linked list of answers.
```

Comment:
Can you elaborate on this? How does setting DGRAM/UDP hint this? Is this linux-specific?

Reply:
It was mainly an empirical observation. I specify only the node and not the service, so specifying the socket_type I noticed that I was getting only one `addrinfo` struct while if leave it as 0 at least linux returns 2 elements SOCK_DGRAM & SOCK_STREAM.

Reply:
I reworked a bit the comment to give more context, let me know if it is less esoteric this way.